### DISCUSSION TITLE
能否新增一个自定义的统一翻译接口

### DISCUSSION THREAD

**Discussion by xingty** [immersive-translate/immersive-translate#602] [Ideas]
### 新功能 | New Feature

1. 在Translation Service中增加一个由扩展提供的统一翻译服务

选中该服务后，需要填写自定义的翻译API服务地址、appKey信息。填写完后，扩展把翻译请求发送到这个API服务地址。

### 为什么您想要添加这些功能 | Why you want to add these features

我是[transcat](https://github.com/xingty/transcat)的开发者，实现了一个翻译服务管理的工具，可以集合任意数量的翻译服务做负载均衡。当前我使用了扩展内的deeplx接口来接入，但我发现扩展好像有意限制了deeplx这个接口的并发量，1s最大好像也就3个请求。所以能否新增一个统一的，不受限制的接口。有这样一个接口也能带来一些好处:

1. 更方便的接入第三方服务
2. 当前众多翻译服务提供商的API大多存在限流，请求并发量大，容易触发限流，需要做负载均衡解决这个问题

### 提议 | Proposal

功能实现我感觉可以大致跟deeplx的接口差不多，只是在它的基础上，再增加一个app_key是最好的。app_key的认证方式，简单一点，可以直接放在header上，但最理想的是使用这个key对request body做一下简单的签名,比如:

```javascript
params = {
  "source_lang": "en",
  "target_lang": "zh",
  "text": "hello,world!"
}
```
那么请求body可能是
```json
{
  "source_lang": "en",
  "target_lang": "zh",
  "text": "hello,world!",
  "timestamp": 1686939101
  "salt": "2a44bfb89a5fe617890b5478fb18f870",
  "sign": "41e0d6b400faee855bfa220b2be0d380"
}
```
签名规则就是把source_lang、target_lang、text、timestamp、salt、app_key按照一定顺序排列，使用hash算法计算出的一个值。

**Comment by theowenyoung** [comment]
很好的提议～ 我们有计划推出一个统一的自定义第三方翻译接口，因为第三方服务总是需要适配沉浸式翻译的语言代码的，所以这个接口定义最好是简单的，统一的标准，对于用户来说，只需要做很小的配置，比如填写一个url即可。

```bash
https://example.com/translate?apiKey=test123
```

API 可能会是这样：

```json
{
  "source_language": "en",
  "target_language": "zh-CN",
  "texts":["sentence1","sentece2"]
}
```
期待的结果：

```json
{
   "translations": [
    {
      "detected_source_language": "EN",
      "text": "Hallo, Welt!"
    }
  ]
}
```

**Reply by xingty** [reply]
非常期待这个接口的加入
